source/fe/fe_system.cc

```diff
@@ -42,8 +42,185 @@ namespace
+
+          unsigned int out_index = 0;
+          for (unsigned int i = 0; i < system_index; ++i)
+            out_index += fe.n_nonzero_components(i);
+          table[base_component][base_index] = out_index;
+        }
+
+    return table;
+  }
+
+  /**
+   * Setup a table of offsets for a nonprimitive FE.
+   */
+  template <int dim, int spacedim = dim>
+  std::vector<typename FESystem<dim, spacedim>::BaseOffsets>
+  setup_nonprimitive_offset_table(const FESystem<dim, spacedim> &fe,
+                                  const unsigned int             base_no)
+  {
+    std::vector<typename FESystem<dim, spacedim>::BaseOffsets> table;
+    const FiniteElement<dim, spacedim> &base_fe = fe.base_element(base_no);
+
+    for (unsigned int system_index = 0; system_index < fe.n_dofs_per_cell();
+         ++system_index)
+      if (fe.system_to_base_index(system_index).first.first == base_no)
+        {
+          const unsigned int base_index =
+            fe.system_to_base_index(system_index).second;
+          Assert(base_index < base_fe.n_dofs_per_cell(), ExcInternalError());
+          table.emplace_back();
+
+          table.back().n_nonzero_components =
+            fe.n_nonzero_components(system_index);
+          auto &in_index = table.back().in_index;
+          in_index       = 0;
+          for (unsigned int i = 0; i < base_index; ++i)
+            in_index += base_fe.n_nonzero_components(i);
+
+          auto &out_index = table.back().out_index;
+          out_index       = 0;
+          for (unsigned int i = 0; i < system_index; ++i)
+            out_index += fe.n_nonzero_components(i);
+        }
+
+    Assert(table.size() ==
+             base_fe.n_dofs_per_cell() * fe.element_multiplicity(base_no),
+           ExcInternalError());
+    return table;
+  }
+
+  /**
+   * Copy data between internal FEValues objects from a primitive FE to the
+   * current FE.
+   */
+  template <int dim, int spacedim = dim>
+  void
+  copy_primitive_base_element_values(
+    const FESystem<dim, spacedim> &fe,
+    const unsigned int             base_no,
+    const unsigned int             n_q_points,
+    const UpdateFlags              base_flags,
+    const Table<2, unsigned int> & base_to_system_table,
+    const FEValuesImplementation::FiniteElementRelatedData<dim, spacedim>
+      &base_data,
+    FEValuesImplementation::FiniteElementRelatedData<dim, spacedim>
+      &output_data)
+  {
+    Assert(fe.base_element(base_no).is_primitive(), ExcInternalError());
+    const unsigned int n_components = fe.element_multiplicity(base_no);
+    const unsigned int n_dofs_per_cell =
+      fe.base_element(base_no).n_dofs_per_cell();
+    for (unsigned int component = 0; component < n_components; ++component)
+      for (unsigned int b = 0; b < n_dofs_per_cell; ++b)
+        {
+          const unsigned int out_index = base_to_system_table[component][b];
+
+          if (base_flags & update_values)
+            for (unsigned int q = 0; q < n_q_points; ++q)
```

Comment:
Would it be faster to swap the ``for`` with the ``if`` and combine the loops?

Reply:
Maybe. I'll try it out.

Reply:
I timed moving switching the flag check and the quadrature point loop and also using `std::copy()` or changing other loop orderings. None of them were faster than the version in the original patch (in fact, they were all pretty close).

Reply:
I think this is the most functions in the `FEValues` context work this way, and I think the code is pretty clean this way. One could reason about whether it is actually necessary to copy around all those arrays, but that is a fundamental design choice of how `FEValues` interacts with `FESystem`. Even though I personally find that inefficient, that ship has sailed, and we have other concepts for higher order shape functions where the cost is significant (matrix-free module).

Reply:
Thank you for trying it out!